mikeio1d/result_network/geometry.py

```diff
@@ -0,0 +1,238 @@
+
+    Parameters
+    ----------
+    x : float
+        X coordinate
+    y : float
+        Y coordinate
+    """
+
+    x: float
+    y: float
+
+    @staticmethod
+    def from_res1d_node(res1d_node) -> NodePoint:
+        """
+        Create a NodePoint from an IRes1DNode object.
+        """
+        xcoord = res1d_node.XCoordinate
+        ycoord = res1d_node.YCoordinate
+        return NodePoint(xcoord, ycoord)
+
+    def to_shapely(self) -> BaseGeometry:
+        return Point(self.x, self.y)
+
+
+@dataclass(frozen=True, order=True)
+class ReachPoint:
+    """
+    A utility class for working with points along a reach.
+
+    Parameters
+    ----------
+    point_type : ReachPointType
+        Either DIGIPOINT or GRIDPOINT
+    chainage : float
+        Chainage along the reach. Need not match geometric distance.
+    x : float
+        X coordinate
+    y : float
+        Y coordinate
+    z : float
+        Z coordinate
+    """
+
+    point_type: ReachPointType = field(compare=False)
+    chainage: float
+    x: float
+    y: float
+    z: float
+
+    def is_digipoint(self):
+        return self.point_type == ReachPointType.DIGIPOINT
+
+    def is_gridpoint(self):
+        return self.point_type == ReachPointType.GRIDPOINT
+
+    def to_shapely(self) -> BaseGeometry:
+        return Point(self.x, self.y)
+
+    @staticmethod
+    def from_digipoint(res1d_digipoint):
+        """
+        Create a ReachPoint from an IRes1DDigiPoint object.
+        """
+        return ReachPoint(
+            ReachPointType.DIGIPOINT,
+            res1d_digipoint.M,
+            res1d_digipoint.X,
+            res1d_digipoint.Y,
+            res1d_digipoint.Z,
+        )
+
+    @staticmethod
+    def from_gridpoint(res1d_gridpoint):
+        """
+        Create a ReachPoint from an IRes1DGridPoint object.
+        """
+        return ReachPoint(
+            ReachPointType.GRIDPOINT,
+            res1d_gridpoint.Chainage,
+            res1d_gridpoint.X,
+            res1d_gridpoint.Y,
+            res1d_gridpoint.Z,
+        )
+
+
+class ReachGeometry:
+    """
+    A utility class for working with reach geometries.
+    """
+
+    def __init__(self, points: List[ReachPoint]):
+        self._points = sorted(points)
+
+    @staticmethod
+    def from_res1d_reaches(res1d_reaches):
+        """
+        Create a ReachGeometry from a list of IRes1DReach objects.
+
+        Parameters
+        ----------
+        res1d_reaches : List[IRes1DReach]
+        """
+
+        if not isinstance(res1d_reaches, Iterable):
+            res1d_reaches = [res1d_reaches]
+
+        points = []
+        for reach in res1d_reaches:
+            points.extend([ReachPoint.from_digipoint(dp) for dp in reach.DigiPoints])
+            points.extend([ReachPoint.from_gridpoint(gp) for gp in reach.GridPoints])
+
+        return ReachGeometry(points)
+
+    @property
+    def chainages(self) -> List[float]:
+        return [p.chainage for p in self._get_unique_points()]
```

Comment:
Does it make sense to store the chainages list instead of recreating it every time or lru_cache takes care of it? Similar question for other properties.

Reply:
Regarding the network structure I think the memory concern will not be very important. But we also saw that it performs good at a larger network (~10000 nodes), so at the end I think no change is needed :-).